File: src/scripts/check_token_optimizer_freshness.py
#!/usr/bin/env python3
"""
Token-Optimizer freshness validator.

Per `road-to-token-optimization.md` P1.3: parses the catalog table inside
`.agent-src.uncondensed/skills/token-optimizer/SKILL.md`, verifies every
cited internal asset exists, and `grep`s the trigger keywords against
each target file. Fails on missing target OR keyword mismatch.

Authoritative-link rows (upstream URLs, planned commands marked TBD) are
recorded but not fetched — they live and die with their upstream and are
out of scope for this CI gate.

Acceptance: stdlib-only, deterministic, exit 0 = clean / exit 1 = drift.
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent.parent
sys.path.insert(0, str(REPO_ROOT / "src" / "scripts"))
from _lib.agent_src import resolve_logical  # noqa: E402

# Post-ADR-017 the source-of-truth lives under whichever package owns
# the skill; resolve_logical() walks every artefact root.
SKILL = resolve_logical("skills/token-optimizer/SKILL.md") or (
    REPO_ROOT / ".agent-src.uncondensed" / "skills" / "token-optimizer" / "SKILL.md"
)

from _lib.agent_src import strip_source_prefix  # noqa: E402

# Catalog row pattern: | name | path | keywords | description |
ROW_RE = re.compile(
    r"^\|\s*`?(?P<name>[^`|]+?)`?\s*\|\s*"
    r"(?P<path>[^|]+?)\s*\|\s*"
    r"(?P<keywords>[^|]+?)\s*\|\s*"
    r"(?P<desc>[^|]+?)\s*\|\s*$"
)
KW_RE = re.compile(r"`([^`]+)`")


def parse_catalog(text: str) -> list[dict[str, str]]:
    rows: list[dict[str, str]] = []
    in_catalog = False
    for line in text.splitlines():
        if line.strip().startswith("## "):
            in_catalog = line.strip() == "## Catalog"
            continue
        if not in_catalog:
            continue
        if line.startswith("|---") or line.startswith("| Asset"):
            continue
        m = ROW_RE.match(line)
        if not m:
            continue
        rows.append(
            {
                "name": m["name"].strip(),
                "path": m["path"].strip(),
                "keywords": m["keywords"].strip(),
                "desc": m["desc"].strip(),
            }
        )
    return rows


def is_external(path: str) -> bool:
    p = path.lower()
    return (
        p.startswith("upstream:")
        or p.startswith("http://")
        or p.startswith("https://")
        or p.startswith("tbd")
        or "github.com" in p
    )


def resolve(path: str) -> Path | None:
    if is_external(path):
        return None
    cleaned = path.strip().lstrip("`").rstrip("`")
    cleaned = cleaned.split(")")[0].lstrip("[(")
    # Catalog rows still cite the legacy .agent-src.uncondensed/ prefix
    # for compactness; resolve those across every packages/* root.
    logical = strip_source_prefix(cleaned)
    if logical is not None:
        hit = resolve_logical(logical)
        if hit is not None:
            return hit
    return (REPO_ROOT / cleaned).resolve()


def check_row(row: dict[str, str]) -> list[str]:
    errs: list[str] = []
    if is_external(row["path"]):
        return errs
    target = resolve(row["path"])
    if target is None or not target.exists():
        errs.append(f"[{row['name']}] target missing: {row['path']}")
        return errs
    body = target.read_text(encoding="utf-8", errors="replace").lower()
    for kw in KW_RE.findall(row["keywords"]):
        kw_lc = kw.strip().lower()
        if not kw_lc:
            continue
        if kw_lc not in body:
            errs.append(
                f"[{row['name']}] trigger keyword '{kw}' not found in "
                f"{row['path']} — catalog row may be stale"
            )
    return errs


def main() -> int:
    if not SKILL.exists():
        print(f"ERROR: token-optimizer skill not found at {SKILL}", file=sys.stderr)
        return 1
    text = SKILL.read_text(encoding="utf-8")
    rows = parse_catalog(text)
    if not rows:
        print(
            "ERROR: token-optimizer SKILL.md has no parseable catalog rows",
            file=sys.stderr,
        )
        return 1
    all_errs: list[str] = []
    checked = 0
    for row in rows:
        errs = check_row(row)
        all_errs.extend(errs)
        if not is_external(row["path"]):
            checked += 1
    print(
        f"token-optimizer freshness: {len(rows)} catalog rows, "
        f"{checked} internal targets checked, {len(all_errs)} drift signal(s)"
    )
    for e in all_errs:
        print(f"  FAIL  {e}")
    return 1 if all_errs else 0


if __name__ == "__main__":
    sys.exit(main())

File: src/scripts/check_token_optimizer_freshness.ts
#!/usr/bin/env tsx
/**
 * Token-Optimizer freshness validator.
 *
 * TypeScript twin of `src/scripts/check_token_optimizer_freshness.py` (ADR-096,
 * Phase 4 / Wave 4c). The CLI contract is mirrored EXACTLY — no flags,
 * exit codes (0 clean, 1 drift / missing skill / unparseable), stdout/stderr
 * split, byte-identical messages, the same catalog-row parsing, the same
 * `resolve()` / `is_external()` logic and the same `_lib/agent_src`
 * resolution. No behaviour changes — latent bugs replicated.
 *
 * Parses the catalog table inside the token-optimizer SKILL.md, verifies
 * every cited internal asset exists, and checks the trigger keywords against
 * each target file. Fails on missing target OR keyword mismatch.
 *
 * Exit codes: 0 = clean, 1 = drift / error.
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

import { resolve_logical, strip_source_prefix } from './_lib/agent_src.js';

const _HERE = fileURLToPath(import.meta.url);
const REPO_ROOT = path.resolve(path.dirname(_HERE), '..', '..');

// Post-ADR-017 the source-of-truth lives under whichever package owns
// the skill; resolve_logical() walks every artefact root.
const SKILL =
    resolve_logical('skills/token-optimizer/SKILL.md') ??
    path.join(REPO_ROOT, '.agent-src.uncondensed', 'skills', 'token-optimizer', 'SKILL.md');

// Catalog row pattern: | name | path | keywords | description |
const ROW_RE =
    /^\|\s*`?(?<name>[^`|]+?)`?\s*\|\s*(?<path>[^|]+?)\s*\|\s*(?<keywords>[^|]+?)\s*\|\s*(?<desc>[^|]+?)\s*\|\s*$/;
const KW_RE = /`([^`]+)`/g;

interface Row {
    name: string;
    path: string;
    keywords: string;
    desc: string;
}

function parse_catalog(text: string): Row[] {
    const rows: Row[] = [];
    let inCatalog = false;
    for (const line of text.split('\n')) {
        if (line.trim().startsWith('## ')) {
            inCatalog = line.trim() === '## Catalog';
            continue;
        }
        if (!inCatalog) {
            continue;
        }
        if (line.startsWith('|---') || line.startsWith('| Asset')) {
            continue;
        }
        const m = ROW_RE.exec(line);
        if (!m) {
            continue;
        }
        const g = m.groups!;
        rows.push({
            name: g['name']!.trim(),
            path: g['path']!.trim(),
            keywords: g['keywords']!.trim(),
            desc: g['desc']!.trim(),
        });
    }
    return rows;
}

function is_external(p: string): boolean {
    const lc = p.toLowerCase();
    return (
        lc.startsWith('upstream:') ||
        lc.startsWith('http://') ||
        lc.startsWith('https://') ||
        lc.startsWith('tbd') ||
        lc.includes('github.com')
    );
}

function resolve(p: string): string | null {
    if (is_external(p)) {
        return null;
    }
    let cleaned = p.trim().replace(/^`+/, '').replace(/`+$/, '');
    // cleaned.split(")")[0].lstrip("[(")
    cleaned = cleaned.split(')')[0]!.replace(/^[[(]+/, '');
    // Catalog rows still cite the legacy .agent-src.uncondensed/ prefix
    // for compactness; resolve those across every packages/* root.
    const logical = strip_source_prefix(cleaned);
    if (logical !== null) {
        const hit = resolve_logical(logical);
        if (hit !== null) {
            return hit;
        }
    }
    return path.resolve(REPO_ROOT, cleaned);
}

function check_row(row: Row): string[] {
    const errs: string[] = [];
    if (is_external(row.path)) {
        return errs;
    }
    const target = resolve(row.path);
    let exists = false;
    if (target !== null) {
        try {
            exists = fs.statSync(target) !== undefined;
        } catch {
            exists = false;
        }
    }
    if (target === null || !exists) {
        errs.push(`[${row.name}] target missing: ${row.path}`);
        return errs;
    }
    const body = fs.readFileSync(target, 'utf-8').toLowerCase();
    KW_RE.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = KW_RE.exec(row.keywords)) !== null) {
        const kwLc = m[1]!.trim().toLowerCase();
        if (!kwLc) {
            continue;
        }
        if (!body.includes(kwLc)) {
            errs.push(
                `[${row.name}] trigger keyword '${m[1]}' not found in ` +
                    `${row.path} — catalog row may be stale`,
            );
        }
    }
    return errs;
}

function main(): number {
    let skillExists = false;
    try {
        skillExists = fs.statSync(SKILL) !== undefined;
    } catch {
        skillExists = false;
    }
    if (!skillExists) {
        process.stderr.write(`ERROR: token-optimizer skill not found at ${SKILL}\n`);
        return 1;
    }
    const text = fs.readFileSync(SKILL, 'utf-8');
    const rows = parse_catalog(text);
    if (rows.length === 0) {
        process.stderr.write('ERROR: token-optimizer SKILL.md has no parseable catalog rows\n');
        return 1;
    }
    const allErrs: string[] = [];
    let checked = 0;
    for (const row of rows) {
        const errs = check_row(row);
        allErrs.push(...errs);
        if (!is_external(row.path)) {
            checked += 1;
        }
    }
    process.stdout.write(
        `token-optimizer freshness: ${rows.length} catalog rows, ` +
            `${checked} internal targets checked, ${allErrs.length} drift signal(s)\n`,
    );
    for (const e of allErrs) {
        process.stdout.write(`  FAIL  ${e}\n`);
    }
    return allErrs.length ? 1 : 0;
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main());
}

export { SKILL, ROW_RE, parse_catalog, is_external, resolve, check_row, main };
